postoffice.8
.\" Copyright (c) 1988, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     %A%
.\"
.Dd November 21, 2004
.Dt POSTOFFICE 8
.Os MASTODON 4
.Sh NAME
.Nm postoffice
.Nd internet mail server
.Sh SYNOPSIS
.Nm postoffice
.Op Ar flags
.Op Ar address ...
.Nm mailq
.Nm newaliases
.Nm runq
.Op Fl v
.Nm smtpd
.Op Ar flags
.Sh DESCRIPTION
.Nm Postoffice
accepts and delivers mail on
.Em TCP/IP
networks. It includes a mail
.Em greylist ,
uses 
.Nm "TCP wrappers"
to restrict client access,
and can be configured to run antivirus
.\" and antispam
software before accepting mail for delivery.
.Pp
With no flags,
.Nm postoffice
reads its standard input
up to an end-of-file
and sends a copy of the message found there
to all of the addresses listed on the command line.
.Pp
Local addresses are looked up in a file
and aliased appropriately.
.Pp
Flags are:
.Bl -tag -width Fl
.It Fl a
Start SMTP servers with session auditing enabled.
If session logging is enabled, 
.Nm postoffice
will log all sorts of information about each
SMTP session to the syslog.  See
.Xr mailaudit 5
for details about the format of the audit records.
.It Fl bd
Run as a daemon.
.Nm Postoffice
will fork and run in background
listening on socket 25 for incoming SMTP
connections.
.It Fl bi
Initialize the alias database.
.It Fl bm
Deliver mail in the usual way (default).
.It Fl bp
Print a listing of the queue.
.It Fl bq
Process saved messages in the queue.
.It Fl bs
Use the SMTP
protocol as described in
.Em RFC821
on standard input and output.
.It Fl C Ns Ar file
Read options from the named configuration file.
.\" .It Fl d Ns Ar X
.\" Set debugging value to
.\" .Ar X .
.It Fl f Ns Ar name
Sets the name of the
.Ql from
person
(i.e., the sender of the mail).
.Fl f
can only be used
by 
.Em root
or if the person you are trying to become
is the same as the person you are.
.It Fl o Ns Ar x Ns = Ns value
Set option
.Ar x
to the specified
.Em value .
Options are described below.
.It Fl q
An alternate form of the
.Fl bd
flag.
.\" .It Fl M Ns Ar ident
.\" Process the queued message with the queue id
.\" .Ar ident .
.\" .It Fl R Ns Ar addr
.\" Process the queued messages that have the string
.\" .Ar addr
.\" in one of the recipient addresses.
.\" .It Fl S Ns Ar addr
.\" Process the queued messages that have the string
.\" .Ar addr
.\" in the sender address.
.It Fl r Ns Ar name
An alternate form of the
.Fl f
flag.
.It Fl v
Go into verbose mode.
.It Fl V
Report version information for this copy of
postoffice, then exit successfully.
.El
.Pp
Various processing options can be set when 
.Nm
is started.
Options may be set either on the command line
using the
.Fl o
flag,
in the configuration file /etc/postoffice.cf, or in a configuration
file specified by the
.Fl C
flag.
The options are:
.Bl -tag -width Fl
.It Li audit
Enable session auditing.
.\" .It Li A Ns Ar file
.\" Use alternate alias file.
.\" .It Li D Ns Ar x
.\" Set the delivery mode to
.\" .Ar x .
.\" Delivery modes are
.\" .Ql i
.\" for interactive (synchronous) delivery,
.\" and
.\" .Ql q
.\" for queue only \- i.e.,
.\" actual delivery is done the next time the queue is run.
.\" .It Li e Ns Ar x
.\" Set error processing to mode
.\" .Ar x .
.\" Valid modes are
.\" .Ql m
.\" to mail back the error message,
.\" and
.\" .Ql q
.\" to throw away error messages
.\" (only exit status is returned).
.\" .It Li cq Ns Ar =N
.\" Checkpoint the queue file after every
.\" .Ar N
.\" successful deliveries (default 10).
.\" This avoids excessive duplicate deliveries
.\" when sending to long mailing lists
.\" interrupted by system crashes.
.\" .It Li queue Ns = Ns Ar queuedir
.\" Select the directory in which to queue messages.
.It Li checkhelo
Check the arguments passed with the 
.Em HELO
and
.Em EHLO
SMTP command and refuse to do an SMTP conversation with the
client if they claim to be the mail server.
.It Li clients Ns = Ns Ar N
Allow only 
.Ar N
clients at a time to connect to the server.
.It Li debug
Enable the ESMTP
.Ql DEBUG
command, which displays various compile and runtime settings,
as well as the current state of the mail envelope.
.It Li delay Ns = Ns Ar N
Set the
.Em greylist Ns ing
delay to
.Ar N
seconds.
The 
.Xr delay
may be specified as a tagged number,
with
.Ql m
being minutes.
and
.Ql h
being hours.
.It Li load Ns = Ns Ar N
When the load average is above
.Ar N ,
.Nm postoffice
will not accept SMTP connections (the default is 4).
.It Li nodaemon
Do not accept mail from
.Em <> .
This can be useful for blocking spammers and viruses, but
it also violates
.%T RFC821 ,
and means that nobody at your site (including mailing lists)
will ever get bounce messages when mail cannot be delivered.
If your machine is connected to a public network,
.Em DON'T
.Em SET
.Em THIS
.Em OPTION
unless you're under attack from a malicious spammer.
.It Li paranoid
Do not accept connections from clients that are not registered in
the 
.Dv DNS .
.It Li hops Ns = Ns Ar N
Set the hop count to
.Ar N .
The hop count is incremented every time the mail is
processed.
When it reaches this limit,
the mail is returned with an error message,
.It Li localmx
Treat connections from sites that use this machine as an MX
as if they were local, and give them the permissions that
local clients have.
.It Li relay
Allow all clients to relay mail through this
server.   If your machine is connected to a public network,
.Em DON'T
.Em SET
.Em THIS
.Em OPTION ,
unless you really want to end up on every antispam blacklist on
the planet.
.It Li relay-host Ns = Ns Ar host
Send all mail destined for remote hosts via this
.Ar host .
The superuser is the only user that can use this option on the
command line or in a configuration file other than
.Pa /etc/postoffice.cf
.It Li size Ns = Ns Ar size
Sets the 
largest message size that 
.Nm postoffice
will accept.
.Xr size
is given as a tagged number,
with
.Ql k
being kilobytes,
and
.Ql m
being megabytes.
.It Li timeout Ns = Ns Ar timeout
Drop the connection to a client if they are idle
longer than
.Ar timeout .
This option violates the word (if not the intent) of the SMTP
specification,
so the timeout should probably be fairly large.
The 
.Xr timeout
may be specified as a tagged number,
with
.Ql m
being minutes,
.Ql h
being hours,
and
.Ql d
being days.
For example,
.Ql timeout=60m
or 
.Ql timeout=1h
both set the timeout to one hour.
.\" .It Li stats Ns Ar file
.\" Save statistics in the named file.
.\" .It Li return Ns Ar time
.\" Set the timeout on undelivered messages in the queue to the specified time.
.\" After delivery has failed
.\" (e.g., because of a host being down)
.\" for this amount of time,
.\" failed messages will be returned to the sender.
.\" The default is three days.
.\" .It Li U Ns Ar userdatabase
.\" If set, a user database is consulted to get forwarding information.
.\" You can consider this an adjunct to the aliasing mechanism,
.\" except that the database is intended to be distributed;
.\" aliases are local to a particular host.
.\" This may not be available if your postoffice does not have the
.\" .Dv USERDB
.\" option compiled in.
.\" .It Li u Ns Ar N
.\" Set the default user id for mailers.
.El
.Pp
In aliases,
the first character of a name may be
a vertical bar to cause interpretation of
the rest of the name as a command
to pipe the mail to.
It may be necessary to quote the name
to keep
.Nm postoffice
from suppressing the blanks from between arguments.
For example, a common alias is:
.Pp
.Bd -literal -offset indent -compact
msgs: "|/usr/bin/msgs -s"
.Ed
.Pp
.Nm postoffice
returns an exit status
describing what it did.
The codes are defined in
.Aq Pa sysexits.h :
.Bl -tag -width EX_UNAVAILABLE -compact -offset indent
.It Dv EX_OK
Successful completion on all addresses.
.It Dv EX_NOUSER
User name not recognized.
.It Dv EX_UNAVAILABLE
Catchall meaning necessary resources
were not available.
.It Dv EX_SYNTAX
Syntax error in address.
.It Dv EX_SOFTWARE
Internal software error,
including bad arguments.
.It Dv EX_OSERR
Temporary operating system error,
such as
.Dq cannot fork .
.It Dv EX_NOPERM
You do not have sufficient permissions.
.It Dv EX_IOERR
.Nm postoffice
was unable to create or write a file.
.El
.Pp
If invoked as
.Nm newaliases ,
.Nm postoffice
will rebuild the alias database.
If invoked as
.Nm mailq ,
.Nm postoffice
will print the contents of the mail queue.
If invoked as
.Nm runq ,
.Nm postoffice
will run through the mail queue and make what deliveries are possible.
If invoked as
.Nm smtpd ,
.Nm postoffice
runs as a SMTP daemon.
.Sh FILES
.Pp
.Bl -tag -width /var/db/smtpauth.db -compact
.It Pa /etc/postoffice.cf
Default
.Nm
options.
.It Pa /etc/aliases
raw data for alias names
.It Pa /etc/aliases.db
data base of alias names
.It Pa /var/log/mail.stat
collected statistics
.It Pa /var/spool/mqueue/*
temp files
.It Pa /var/db/smtpauth.db
.Nm greylist
database.
.El
.Sh SEE ALSO
.Xr mail 1 ,
.Xr rmail 1 ,
.Xr syslog 3 ,
.Xr aliases 5 ,
.Xr mailaudit 5 ,
.Xr smtpauth 5 ,
.Xr mailaddr 7 ,
.Xr authexpire 8 ,
.Xr sendmail 8 ,
.Xr smtpd 8 ;
.Pp
DARPA
Internet Request For Comments
.%T RFC819 ,
.%T RFC821 ,
.%T RFC822 .


postoffice.c
/*
 * smtp gateway server
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <syslog.h>
#include <sys/utsname.h>
#include <netdb.h>
#include <sysexits.h>

#if HAVE_LIBGEN_H
#include <libgen.h>
#endif

extern char myversion[];

#include "letter.h"
#include "smtp.h"


#ifndef HAVE_SETPROCTITLE
char *argv0;
int   szargv0;
#endif


extern int   z_getopt();
extern char* z_optarg;
extern int   z_optind;

void
superpowers()
{
    if (getuid() != geteuid()) {
	/* pick up our superpowers (if any) */
	setuid(geteuid());
	setgid(getegid());
    }
}

char *pgm;

float
main(int argc, char **argv)
{
    int opt;
    extern struct in_addr *local_if_list();
    struct sockaddr_in *peer = 0;	/* peer for -bs (for debugging) */
    static ENV env;
    char *from = 0;
    char *options = "aA:B:C:dF:f:r:b:o:h:R:q;GUVimnv";
    char *modes = "sqpmid";
    struct utsname sys;
    struct hostent *p;
    int debug = 0;
    int sendmaild0 = 0;
    int qruntime = 0;

    env.bmode = 'm';
    env.local_if = local_if_list();
    env.relay_ok = 1;		/* allow local relaying */
    env.verbose = 0;		/* be chattery */
    env.verify_from = 1;	/* verify the host of MAIL FROM:<user@host> */
    env.nodaemon = 0;		/* allow MAIL FROM:<> */
    env.delay = 3600;		/* greylist delay */
    env.largest = 0;
    env.debug = 0;
    env.sender = getuid();
    env.timeout = 300;
    env.qreturn = 86400*3;	/* 3 days */
    env.max_loadavg = 4.0;
#if HAVE_STATFS || HAVE_STATVFS
    env.minfree = 10*1000*1024;	/* 10m free for messages */
#else
    env.minfree = 0;
#endif
    env.max_clients = 100;	/* should be fairly ridiculous */
    env.max_hops = 100;		/* (ditto) */

#ifndef HAVE_SETPROCTITLE
    argv0 = argv[0];
    szargv0 = 80;
    /*for (env.szargv0=i=0; i<argc; i++)
	env.szargv0 += strlen(argv[i]) + 1;*/
#endif

    if ( p = gethostbyname((uname(&sys) == 0) ? sys.nodename : "localhost") )
	env.localhost = strdup(p->h_name);
    else
	env.localhost = "localhost";

#if HAVE_BASENAME
    pgm = strdup(basename(argv[0]));
#else
    {   char *avp = strrchr(argv[0], '/');
	pgm = strdup(avp ? avp+1 : argv[0]);
    }
#endif
    openlog(pgm, LOG_PID, LOG_MAIL);

#if 0
    {	int i;
	for (i=0; i < argc; i++)
	    syslog(LOG_INFO, "argv[%d] = %s", i, argv[i]);
    }
#endif

#define SAME(a,b)	(strcasecmp(a,b) == 0)

    /* handle magic program names */
    if ( SAME(pgm, "mailq") ) {
	options = "vV";
	env.bmode = 'p';
    }
    else if ( SAME(pgm, "sendmail") || SAME(pgm, "send-mail") ) {
	options = "A:b:d;F:f:imo:r:q;Vvt";
	env.bmode = 'm';
    }
    else if ( SAME(pgm, "runq") ) {
	options = "o:Vv";
	env.bmode = 'q';
    }
    else if ( SAME(pgm, "newaliases") ) {
	options = "vV";
	env.bmode = 'i';
    }
    else if ( SAME(pgm, "smtpd") ) {
	options = "aA:B:C:do:h:q:R:GUmnvV";
	env.bmode = 'd';
    }

    while ( (opt = z_getopt(argc, argv, options)) != EOF) {
	switch (opt) {
	case 'a':
		env.auditing = 1;
		break;
	case 'C':
		if (configfile( z_optarg, &env ) == 0) {
		    perror(z_optarg);
		    exit(EX_NOINPUT);
		}
		break;
	case 'R':
#ifdef USE_PEER_FLAG
		{   struct sockaddr_in fake;
		    long ip;

		    if ( (ip = inet_addr(z_optarg)) != -1) {
			fake.sin_addr = inet_makeaddr(ntohl(ip), 0L);
			fake.sin_family = AF_INET;
			peer = &fake;
			env.relay_ok = islocalhost(&env, &fake.sin_addr);
		    }
		}
#endif
		break;
	case 'q':
		if (z_optarg && *z_optarg) {
		    value(z_optarg, &qruntime, "m=1,h=60,d=1440");
		}
		else
		    env.bmode = 'q';
		break;
	case 't':
		env.trawl = 1;	/* scrounge the message header looking for
				 * recipients (pine sucks)
				 */
		break;
	case 'r':
	case 'f':
		env.forged = (env.sender != 0);
		from = z_optarg;
		break;
	case 'd':
		if (z_optarg)
		    sendmaild0 = (strcmp(z_optarg, "0") == 0)
			      || (strncmp(z_optarg, "0.", 2) == 0);
		else
		    debug = 1;
		break;
	case 'b':
		env.bmode = z_optarg[0];
		break;
	case 'o':
		set_option( z_optarg, &env );
		break;
	case 'v':
		env.verbose = 1;
		break;
	case 'V':
		if (strcasecmp(pgm, "postoffice") == 0)
		    printf("%s %s\n", pgm, myversion);
		else
		    printf("%s: postoffice %s\n", pgm, myversion);
		exit(EX_OK);
	}
    }

    if (sendmaild0)
	printf("Version %s\n", myversion);

    if (strchr(modes, env.bmode)) {
	switch (env.bmode) {
	case 's':
		superpowers();
		configfile("/etc/postoffice.cf", &env);
		smtp(stdin, stdout, peer, &env);
		exit(EX_TEMPFAIL);
	case 'd':
		if (getuid() == 0) {
		    daemonize(&env, debug);
		    if (qruntime) {
			pid_t qd;

			if ( (qd=fork()) == -1) {
			    syslog(LOG_ERR, "starting runqd: %m");
			    exit(EX_TEMPFAIL);
			}
			else if (qd == 0) {
			    runqd(&env,qruntime);
			    exit(EX_TEMPFAIL);
			}
		    }

		    configfile("/etc/postoffice.cf", &env);
		    if (env.auditing)
			auditon();
		    else
			auditoff();
		    server(&env, debug);
		    exit(EX_OK);
		}
		fprintf(stderr, "%s: Permission denied.\n", pgm);
		exit(EX_NOPERM);
	case 'p':
		superpowers();
		listq();
		break;
	case 'm':
		if (qruntime) {
		    if (getuid() == 0) {
			daemonize(&env, debug);
			runqd(&env, qruntime);
			exit(EX_OK);
		    }
		    fprintf(stderr, "%s: Permission denied.\n", pgm);
		    exit(EX_NOPERM);
		}
		else {
		    superpowers();
		    configfile("/etc/postoffice.cf", &env);
		    mail(from, argc-z_optind, argv+z_optind, &env);
		}
		break;
	case 'q':
		superpowers();
		configfile("/etc/postoffice.cf", &env);
		runq(&env);
		break;
	case 'i':	/* initialize alias database */
		superpowers();
		newaliases(argc-z_optind, argv+z_optind);
		break;
	}
	exit(EX_OK);
    }
    fprintf(stderr, "%s: mode <%c> is not implemented\n", pgm, env.bmode);
    exit(EX_USAGE);
}
